include/gridtools/boundary_conditions/apply_gpu.hpp

```diff
@@ -26,54 +28,68 @@ namespace gridtools {
+                init<zero_, minus_, minus_>(p);
+                init<zero_, minus_, zero_>(p);
+                init<zero_, minus_, plus_>(p);
 
-                m_values[1][1][0] = p(direction<zero_, zero_, minus_>{});
+                init<zero_, zero_, minus_>(p);
 
-                m_values[1][1][2] = p(direction<zero_, zero_, plus_>{});
+                init<zero_, zero_, plus_>(p);
 
-                m_values[1][2][0] = p(direction<zero_, plus_, minus_>{});
-                m_values[1][2][1] = p(direction<zero_, plus_, zero_>{});
-                m_values[1][2][2] = p(direction<zero_, plus_, plus_>{});
+                init<zero_, plus_, minus_>(p);
+                init<zero_, plus_, zero_>(p);
+                init<zero_, plus_, plus_>(p);
 
-                m_values[2][0][0] = p(direction<plus_, minus_, minus_>{});
-                m_values[2][0][1] = p(direction<plus_, minus_, zero_>{});
-                m_values[2][0][2] = p(direction<plus_, minus_, plus_>{});
+                init<plus_, minus_, minus_>(p);
+                init<plus_, minus_, zero_>(p);
+                init<plus_, minus_, plus_>(p);
 
-                m_values[2][1][0] = p(direction<plus_, zero_, minus_>{});
-                m_values[2][1][1] = p(direction<plus_, zero_, zero_>{});
-                m_values[2][1][2] = p(direction<plus_, zero_, plus_>{});
+                init<plus_, zero_, minus_>(p);
+                init<plus_, zero_, zero_>(p);
+                init<plus_, zero_, plus_>(p);
 
-                m_values[2][2][0] = p(direction<plus_, plus_, minus_>{});
-                m_values[2][2][1] = p(direction<plus_, plus_, zero_>{});
-                m_values[2][2][2] = p(direction<plus_, plus_, plus_>{});
+                init<plus_, plus_, minus_>(p);
+                init<plus_, plus_, zero_>(p);
+                init<plus_, plus_, plus_>(p);
             }
 
-            GT_FUNCTION
             precomputed_pred(precomputed_pred const &) = default;
 
             template <gridtools::sign I, gridtools::sign J, gridtools::sign K>
             GT_FUNCTION bool operator()(direction<I, J, K>) const {
-                return m_values[static_cast<int>(I) + 1][static_cast<int>(J) + 1][static_cast<int>(K) + 1];
+                static constexpr uint_t index = direction_index<I, J, K>();
+                return (m_values >> index) & 0x1;
             }
+
+          private:
+            template <gridtools::sign I, gridtools::sign J, gridtools::sign K, class Predicate>
+            void init(Predicate const &p) {
+                static constexpr uint_t mask = 0x1 << direction_index<I, J, K>();
+                m_values = (m_values & ~mask) | (p(direction<I, J, K>{}) ? mask : 0);
+            }
+
+            template <gridtools::sign I, gridtools::sign J, gridtools::sign K>
+            GT_FUNCTION static constexpr uint_t direction_index() {
+                return (static_cast<int>(I) + 1) * 9 + (static_cast<int>(J) + 1) * 3 + (static_cast<int>(K) + 1);
```

Comment:
I am not a friend of magic numbers, can you 9 -> stride_i and 3 -> stride_j or something like that?